momo_modules/gsm_module/src/app_main.c

```diff
@@ -6,57 +6,111 @@
 #include "mib12_api.h"
 #include "watchdog.h"
 #include "gsm_serial.h"
+#include "gsm.h"
+#include "gprs.h"
+#include "http.h"
 #include "simcard.h"
-#include "gsm_module.h"
 #include "global_state.h"
-#include "gsm_strings.h"
-#include "timer1.h"
 #include "port.h"
+#include "intel_hex.h"
 
-#define SHUTDOWN_TIMEOUT 240 //One minute, in half seconds
-
-static void report_result(void)
+static void report_result(bool success)
 {
 	bus_master_begin_rpc();
 	mib_packet.param_spec = plist_empty();
 	mib_packet.feature = 60;
-	mib_packet.command = ( cmgs_matched() )? 0xF0 : 0xF1;
+	mib_packet.command = ( success )? 0xF0 : 0xF1;
 	bus_master_send_rpc(8);
 }
 
 static void capture_error(void)
 {
-	while ( gsm_receiveone() == 0 )
-		;
+	uint8 len = gsm_read( mib_buffer, kBusMaxMessageSize );
+	if ( len == 0 )
+		return;
 
 	bus_master_begin_rpc();
-	bus_master_prepare_rpc( 42, 0x20, plist_with_buffer( 0, copy_to_mib() ) );
+	bus_master_prepare_rpc( 42, 0x20, plist_with_buffer( 0, len ) );
 	bus_master_send_rpc( 8 );
 }
 
+uint8 atoi8( const char* str )
+{
+	uint8 i = 0;
+	while ( *str != 0 )
+	{
+		i *= 10;
+		i += *str - '0';
+	}
+	return i;
+}
+
+extern uint8 response_status;
+extern char* uint_buf;
+bool http_await_response() {
+	gsm_expect( "+HTTPACTION:" );
+	char* ptr = uint_buf;
+	if ( gsm_await( 1 ) != 1 )
+		return false;
+	gsm_rx(); // '1', '2', or '3'
+	gsm_rx(); // ','
+	while ( gsm_rx() )
+	{
+		if ( gsm_rx_peek() == ',' )
+			break;
+		*(ptr++) = gsm_rx_peek();
+	}
+	response_status = atoi8( uint_buf );
+	while ( gsm_rx() )
+	{
+		if ( gsm_rx_peek() == '\r' )
+			break;
+	}
+	return true;
+}
+
 void task(void)
 {
 	wdt_disable();
 	
 	//Don't sleep while the module's on so that we don't miss a
 	//serial message
-	uint8 timeout_counter = SHUTDOWN_TIMEOUT;
 	while(state.module_on)
 	{
-		if ( gsm_receiveone() == 0 )
-			timeout_counter = SHUTDOWN_TIMEOUT;
-		else if ( state.shutdown_pending )
-			--timeout_counter; // decrement if we timed out waiting for an RX byte.
+		gsm_rx();
 
 		if ( state.shutdown_pending )
 		{
-			if ( timeout_counter == 0 || cmgs_matched() || err_matched() )
+			uint8 result, timeout_seconds = 8*60;
+			if ( state.stream_type == kStreamSMS )
+			{
+				gsm_expect( "+CMGS:" );
+				gsm_expect2( "ERROR" );
```

Comment:
`expect` tells the parser to check for one value, if you then call `expect2` it will also check for the second value and let you know which one was found.  I can consolidate them into one function since that will likely be easier to understand.